examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,410 @@
+    }
+
+    // -- Cleanup
+    PetscCall(DMSwarmRestoreField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coord_points));
+    PetscCall(DMSwarmRestoreField(dm_swarm, "u", NULL, NULL, (void **)&u_points_array));
+    PetscCall(DMSwarmSortRestoreAccess(dm_swarm));
+    PetscCall(DMRestoreLocalVector(dm_mesh, &U_loc));
+  }
+
+  // Interpolate from mesh to points via libCEED
+  {
+    IS                 is_points;
+    Vec                U_loc, ref_coords;
+    PetscInt           num_points_local;
+    const PetscInt    *all_points;
+    PetscInt           cell_start, cell_end;
+    PetscScalar       *u_points_array;
+    const PetscScalar *coords_points_ref;
+    CeedVector         u_evec, u_cell;
+
+    // -- Get mesh values
+    PetscCall(DMGetLocalVector(dm_mesh, &U_loc));
+    PetscCall(DMGlobalToLocal(dm_mesh, U_mesh, INSERT_VALUES, U_loc));
+    {
+      const PetscScalar *u_array;
+      CeedVector         u_lvec;
+
+      PetscCall(VecGetArrayRead(U_loc, &u_array));
+      CeedElemRestrictionCreateVector(restriction_mesh_u, &u_lvec, &u_evec);
+      CeedVectorSetArray(u_lvec, CEED_USE_POINTER, CEED_MEM_HOST, (CeedScalar *)u_array);
+      CeedElemRestrictionApply(restriction_mesh_u, CEED_NOTRANSPOSE, u_lvec, u_evec, CEED_REQUEST_IMMEDIATE);
+      CeedVectorTakeArray(u_lvec, CEED_MEM_HOST, (CeedScalar **)&u_array);
+      PetscCall(VecRestoreArrayRead(U_loc, &u_array));
+      CeedVectorDestroy(&u_lvec);
+    }
+    {
+      CeedInt P;
+
+      CeedBasisGetNumNodes(basis_mesh_u, &P);
+      CeedVectorCreate(ceed, P, &u_cell);
+    }
+
+    // -- Get swarm values
+    PetscCall(DMPlexGetHeightStratum(dm_mesh, 0, &cell_start, &cell_end));
+    PetscCall(DMSwarmGetField(dm_swarm, "u", NULL, NULL, (void **)&u_points_array));
+    PetscCall(DMSwarmCreateReferenceCoordinates(dm_swarm, dm_mesh, &is_points, &ref_coords));
+    PetscCall(ISGetSize(is_points, &num_points_local));
+    PetscCall(ISGetIndices(is_points, &all_points));
+    PetscCall(VecGetArrayRead(ref_coords, &coords_points_ref));
+
+    // -- Interpolate values to each swarm point, one element in the background mesh at a time
+    for (PetscInt cell = cell_start; cell < cell_end; cell++) {
+      CeedVector u_points, x_points;
+      PetscInt   num_points_in_cell = 0;
+
+      for (PetscInt i = 0; i < num_points_local; i++) {
+        if (all_points[i] == cell) num_points_in_cell++;
+      }
+      CeedVectorCreate(ceed, num_points_in_cell, &u_points);
+      CeedVectorCreate(ceed, num_points_in_cell * dim, &x_points);
+
+      // ---- Reference coordinates for swarm points in background mesh element
+      {
+        PetscInt    p = 0;
+        CeedScalar *x;
+
+        CeedVectorGetArrayWrite(x_points, CEED_MEM_HOST, &x);
+        for (PetscInt i = 0; i < num_points_local; i++) {
+          if (all_points[i] == cell) {
+            for (PetscInt d = 0; d < dim; d++) x[p * dim + d] = coords_points_ref[i * dim + d];
+            p++;
+          }
+        }
```

Comment:
Need to think more here - I use this pattern since I'm trying to only use the data available from the call to `DMSwarmCreateReferenceCoordinates` (the new function we're creating in this MR)